### ISSUE TITLE
RELAX looping error with likelihood/Tree

### ISSUE THREAD

**Issue by ltornabene** [veg/hyphy#366]
I created a wrapper to loop RELAX over a series of files.  After the first successful analysis, the second analysis gets stuck after calculating likelihood scores of the MG94+GTR model.  

Below is my wrapper:

---

fileToExe = HYPHY_LIB_DIRECTORY + "TemplateBatchFiles" + DIRECTORY_SEPARATOR + "RELAX.bf";

/\* a list of file paths */
SetDialogPrompt ( "Provide a list of files to process:" );
fscanf ( PROMPT_FOR_FILE, "Lines", _filesToProcessUsingBSREL );

fprintf (stdout, "[READ ", Columns (_filesToProcessUsingBSREL), " file path lines]\n");

inputRedirect = {};
inputRedirect["01"]="Universal";
inputRedirect["03"]="Y"; 
inputRedirect["04"]="Foreground";
inputRedirect["05"]="Minimal";

for ( _fileLine = 0; _fileLine < Columns ( _filesToProcessUsingBSREL ); _fileLine += 1 ) {

  inputRedirect [ "02" ]    = _filesToProcessUsingBSREL[ _fileLine ];
  ExecuteAFile ( fileToExe, inputRedirect );

---

HERE IS THE ANALYSIS OUTPUT AND ERROR:
# Analysis Version

1.00

[RELAX] Loaded an MSA with 43 sequences and 50 codons from '/Users/Li_lab/Desktop/test_directory/Danio_rerio.1.18281190.18281339.fas.merged' 

A tree was found in the data file:
((42SAFA{Background}:0.05267092626752071,43ECBI{Background}:0.059719446249721884){Background}:0.08854695406059913,((39KUGU{Background}:0.050992642344672226,((40SPOR{Background}:0.016128443041980352,41PTKA{Background}:0.01767719235779238){Background}:0.05437241489941111,(((04TYPA{Foreground}:0.042013938281525345,05MIVE{Foreground}:0.021423553896697672){Foreground}:0.025210168345535553,((07ELAC{Background}:0.041759450893116154,(08DOMA{Background}:0.04430149847283252,06GODO{Background}:0.03289225939843442){Background}:0.005019149098097652){Background}:0.06358496757224394,((09BUKO{Background}:0.06936653618968926,(10KRNA{Background}:0.044852620130474016,11OXMA{Background}:0.019567633203834832){Background}:0.00517371209306087){Background}:0.019567341315400163,((19STNE{Background}:0.07297635383125523,((12AMYA{Background}:0.005279134118908592,13AMWH{Background}:0.006928394611363866){Background}:0.04024235988779685,(20MIDO{Background}:0.08886505088279857,(14POMI{Background}:0.04889211071309922,((17VAPU{Background}:0.01390849361649808,18VAST{Background}:0.017299054766840467){Background}:0.0645606808306349,16GORO{Background}:0.09990016443392162){Background}:0.004875467301615783){Background}:0.003567365141291836){Background}:0.003191128353391436){Background}:0.004118819166192574){Background}:0.035395687977535416,(((((22ACOM{Background}:0.0059181825223686545,21ACFL{Background}:0.0037868625156202134){Background}:0.019173679799320678,30LOOC{Background}:0.02129609515238016){Background}:0.021072872856316576,((28RHGI{Background}:0.01906664446433621,29RHDA{Background}:0.013318975251901123){Background}:0.017092213780554085,(26TRBA{Background}:0.010131467337090402,27TRBI{Background}:0.0123355033638408){Background}:0.023497631043513587){Background}:0.0030927585710387897){Background}:0.009767203822340555,(24MUCA{Background}:0.012422177386830742,25MUAB{Background}:0.014610550746749907){Background}:0.06070608997614041){Background}:0.008865306275435974,(((37ODLA{Background}:0.034537893495984455,(34BOPE{Background}:0.021444996785266488,(36TRVA{Background}:0.017805743225470492,35TASP{Background}:0.013598535081434658){Background}:0.006104097406007018){Background}:0.0019848083948234103){Background}:0.019709830223911602,(32PEKA{Background}:0.017268869967334066,33PEMO{Background}:0.01913784106483372){Background}:0.02227400330768635){Background}:0.014442382586182946,(38SICR{Background}:0.0599181753886573,23CTBO{Background}:0.10412514631435152){Background}:0.027779630305049003){Background}:0.009215840604451225){Background}:0.03617977368850973){Background}:0.05016727264035682){Background}:0.012867761920640364){Background}:0.01124361054193743){Background}:0.021381601976459396,(01RHAS{Background}:0.029548198778566778,(02ODPO{Background}:0.029357919684954513,03MISW{Background}:0.0366399945843457){Background}:0.010941265651113178){Background}:0.025407735931666956){Background}:0.10536530026422872){Background}:0.008561061094241363){Background}:0.04072136180007481,44SICH{Background}:0.04088658831193961){Background}:0.019583615562667012,46QUNI{Background}:0.08607944570557631)

Would you like to use it:(Y/N)?

[RELAX] Selected 3 branches as the test set: 04TYPA,05MIVE,Node13 
[RELAX] Obtaining branch lengths under the GTR model 
[RELAX] Log(L) = -1595.513175782655 
[RELAX] Obtaining omega and branch length estimates under the partitioned MG94xGTR model 
[RELAX] Log(L) = -1362.611424880734 
Error:
RELAX.tree is linked to a likelihood function (relax.LF) and cannot be modified  in call to SetParameter(

Function call stack
1 : Set parameter MODEL of RELAX.tree.04TYPA to RELAX.test_model
    Standard input redirect:
        Empty Associative List-------
2 : ExecuteCommands in string "SetParameter (`id`."+model.applyModelToTree.list[model.applyModelToTree.b]+",MODEL,"+model.applyModelToTree.apply_model+")" using basepath /usr/local/lib/hyphy/TemplateBatchFiles/lib2014/models/.
    Standard input redirect:
        Empty Associative List-------
3 : model.applyModelToTree("RELAX.tree",relax.tree,RELAX.model_assignment,relax.selected_branches)
    Standard input redirect:
        Empty Associative List-------
4 : ExecuteAFile from file fileToExe using basepath /Users/Li_lab/Desktop/Selection_Pipeline_Scripts/. reading input from inputRedirect
{"01":"Universal",
"03":"Y",
"04":"Foreground",
"05":"Minimal",
"02":"/Users/Li_lab/Desktop/test_directory/Danio_rerio.1.18281190.18281339.fas.merged"}

---

hyphymp(34251) malloc: **\* error for object 0x7fd56b85bbf0: pointer being freed was not allocated
**\* set a breakpoint in malloc_error_break to debug
Abort trap: 6

Sorry for keeping you guys busy with all my question!

Luke

**Comment by spond** [comment]
Hi @ltornabene,

This is a re-entrancy bug. I've never tested the RELAX.bf file in this context, and since it uses a lot of new libraries, which haven't been fully troubleshot for issues like this. I'll take a look early next week and see what I can do about fixing it. In the meantime, try adding the line

```
DeleteObject (relax.LF);
```

right after you call `ExecuteAFile` in the wrapper and see if that helps.

Sergei

**Comment by ltornabene** [comment]
Hi Sergei, 
I did try removing the 'relax.LF' in the loop.  Removing it did allow the loop to pass to the second dataset, but with strange results.  Specifically, I removed relax.lf and I looped RELAX over the same alignment/tree file 3 times.  I also replicated this entire looping excercise in four separate terminal windows. In all four looping exercises the first alignment ran very quickly, but the second and third alignments each took more than fives times longer than the first.  Most alarmingly, the likelihood scores for the null and laternative models for each consecutive loop were very different (~ 10.0 points higher each loop).  This was consisent across the 4 replicate analyses, i.e. all four analyses of the first alignments had similar scores, as did all four analyses of second alignments, etc. Thus, there is a  consistent predictable difference in likelihood scores and computing time from removing 'relax.lf', and it may not be the fix we are looking for.
